packages/realm/src/schema/normalize.ts

```diff
@@ -43,163 +50,430 @@ export const PRIMITIVE_TYPES = new Set<PropertyTypeName>([
   "uuid",
 ]);
 
-function isPrimitive(type: string | undefined) {
-  return PRIMITIVE_TYPES.has(type as PropertyTypeName);
-}
+const COLLECTION_TYPES = new Set<CollectionPropertyTypeName>(["list", "dictionary", "set"]);
 
-function validateCanonicalPropertySchema(
-  objectSchemaName: string,
-  { name, type, objectType, optional }: CanonicalObjectSchemaProperty,
-) {
-  if (type === "list" && objectType === "list") {
-    throw new Error(`List property '${objectSchemaName}#${name}' cannot have list elements`);
-  }
-  if (type === "list" && !isPrimitive(objectType) && optional) {
-    throw new Error(`List property '${objectSchemaName}#${name}' of '${objectType}' elements, cannot be optional`);
-  }
-  if (objectType === "") {
-    throw new Error(`Property '${objectSchemaName}#${name}' cannot have an empty object type`);
-  }
+const COLLECTION_SHORTHAND_TO_NAME: Readonly<Record<string, string>> = {
+  "[]": "list",
+  "{}": "dictionary",
+  "<>": "set",
+};
+
+function isPrimitive(type: string | undefined): boolean {
+  return PRIMITIVE_TYPES.has(type as PrimitivePropertyTypeName);
 }
 
-export const COLLECTION_TYPES: PropertyTypeName[] = ["set", "dictionary", "list"];
+function isCollection(type: string | undefined): boolean {
+  return COLLECTION_TYPES.has(type as CollectionPropertyTypeName);
+}
 
-function removeUndefinedValues<T extends Record<string, unknown>>(obj: T): T {
-  return Object.fromEntries(Object.entries(obj).filter(([, v]) => typeof v !== "undefined")) as T;
+function isUserDefined(type: string | undefined): boolean {
```

Comment:
Good idea! That will definitely improve UX and prevent typos from being interpreted as a user-defined type.

Reply:
Update: Unfortunately that suggestion cannot currently be implemented due to `Realm.createTemplateObject()` since it does not have access to the object schema names (from the realm schema) which it would have needed to pass to `normalizeObjectSchema`. Mathias will look into the necessity of the `createTemplateObject` feature and if it still needs to be supported.